File: leetcode/hard/315.py
# https://leetcode.com/problems/count-of-smaller-numbers-after-self/

import unittest
from typing import List


class Solution:
    def findIndexToAdd(self, sorted_nums: List[int], n: int) -> int:
        if len(sorted_nums) == 0:
            return 0

        left, right = 0, len(sorted_nums)-1
        while left <= right:
            mid = (left + right) // 2
            if n > sorted_nums[mid]:
                left = mid + 1
            else:
                right = mid - 1

        return max(left, right)

    def countSmaller(self, nums: List[int]) -> List[int]:
        nums.reverse()

        sorted_nums = []
        indices = []

        for num in nums:
            index = self.findIndexToAdd(sorted_nums, num)
            sorted_nums.insert(index, num)
            indices.append(index)

        return indices[::-1]


class TestSolution(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()
    
    def testCountSmaller(self):
        test_cases = [
            [[5,2,6,1],[2,1,1,0]],
            [[-1],[0]],
            [[-1,-1],[0,0]]
        ]

        for test_case in test_cases:
            nums, expected = test_case
            answer = self.solution.countSmaller(nums)
            self.assertEqual(answer, expected)


if __name__ == "__main__":
    unittest.main()

File: leetcode/hard/315.ts
// https://leetcode.com/problems/count-of-smaller-numbers-after-self/

function countSmaller(nums: number[]): number[] {
    function findIndexToAdd(sortedNums: number[], n: number) {
        if (sortedNums.length === 0) {
            return 0;
        }
        let left = 0, right = sortedNums.length - 1;
        while (left <= right) {
            const mid = Math.floor((left + right) / 2);
            if (sortedNums[mid] < n) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return Math.max(left, right);
    }

    nums.reverse();

    let sortedNums: number[] = [];
    const indices = [];

    for (let i = 0; i < nums.length; i++) {
        const index = findIndexToAdd(sortedNums, nums[i]);
        sortedNums.splice(index, 0, nums[i]);
        indices.push(index);
    }

    indices.reverse();
    return indices;
};

const testCases = [
    [[5, 2, 6, 1], [2, 1, 1, 0]],
    [[-1], [0]],
    [[-1, -1], [0, 0]],
    [[1, 2, 3, 4], [0, 0, 0, 0]],
    [[4, 3, 2, 1], [3, 2, 1, 0]]
];

for (const testCase of testCases) {
    const [nums, expected] = testCase;
    console.log(countSmaller(nums));
}

export { };